File: benchmarks/comparison/hof_pipeline.rs
fn double_val(x: i64) -> i64 { x * 2 }
fn add_one(x: i64) -> i64 { x + 1 }
fn apply(f: fn(i64) -> i64, x: i64) -> i64 { f(x) }

fn main() {
    let mut total: i64 = 0;
    for i in 0i64..10_000_000 {
        total += apply(double_val, i);
        total += add_one(double_val(i));
    }
    println!("{}", total);
}

File: benchmarks/comparison/hof_pipeline.c
#include <stdio.h>
#include <stdint.h>

typedef int64_t (*fn_i64)(int64_t);

static int64_t double_val(int64_t x) { return x * 2; }
static int64_t add_one(int64_t x) { return x + 1; }

static int64_t apply(fn_i64 f, int64_t x) { return f(x); }

int main(void) {
    int64_t total = 0;
    for (int64_t i = 0; i < 2000000000; i++) {
        total += apply(double_val, i);
        total += add_one(double_val(i));
        total ^= i;
    }
    printf("%ld\n", total);
    return 0;
}
